tools/webtransport/h3/webtransport_h3_server.py

```diff
@@ -0,0 +1,307 @@
+        return stream_id
+
+    def send_stream_data(self,
+                         stream_id: int,
+                         data: bytes,
+                         end_stream: bool = False) -> None:
+        """
+        Send data on the specific stream.
+
+        :param stream_id: The stream ID on which to send the data.
+        :param data: The data to send.
+        :param end_stream: If set to True, the stream will be closed.
+        """
+        self._http._quic.send_stream_data(stream_id=stream_id,
+                                          data=data,
+                                          end_stream=end_stream)
+
+    def send_datagram(self, data: bytes) -> None:
+        """
+        Send data using a datagram frame.
+
+        :param data: The data to send.
+        """
+        self._http.send_datagram(flow_id=self.session_id, data=data)
+
+
+class WebTransportEventHandler:
+    def __init__(self, session: WebTransportSession,
+                 callbacks: Dict[str, Any]) -> None:
+        self._session = session
+        self._callbacks = callbacks
+
+    def _run_callback(self, callback_name: str, *args, **kwargs) -> None:
+        if callback_name not in self._callbacks:
+            return
+        try:
+            self._callbacks[callback_name](*args, **kwargs)
+        except Exception as e:
+            _logger.warn(str(e))
+            traceback.print_exc()
+
+    def connect_received(self, response_headers: List[Tuple[bytes,
+                                                            bytes]]) -> None:
+        self._run_callback("connect_received", self._session.request_headers,
+                           response_headers)
+
+    def session_established(self) -> None:
+        self._run_callback("session_established", self._session)
+
+    def stream_data_received(self, stream_id: int, data: bytes,
+                             stream_ended: bool) -> None:
+        self._run_callback("stream_data_received", self._session, stream_id,
+                           data, stream_ended)
+
+    def datagram_received(self, data: bytes) -> None:
+        self._run_callback("datagram_received", self._session, data)
+
+
+class SessionTicketStore:
+    """
+    Simple in-memory store for session tickets.
+    """
+    def __init__(self) -> None:
+        self.tickets: Dict[bytes, SessionTicket] = {}
+
+    def add(self, ticket: SessionTicket) -> None:
+        self.tickets[ticket.ticket] = ticket
+
+    def pop(self, label: bytes) -> Optional[SessionTicket]:
+        return self.tickets.pop(label, None)
+
+
+# TODO(bashi): Consider adding more configuration information, such as
+# providing access to StashServer.
+class WebTransportH3Server:
+    """
+    A WebTransport over HTTP/3 for testing.
+
+    :param host: Host from which to serve.
+    :param port: Port from which to serve.
+    :param doc_root: Document root for serving handlers.
+    :param cert_path: Path to certificate file to use.
```

Comment:
I'm not sure if all of these bits will be required still if you change the way the server is started to threaded. If it isn't (I'm really not sure) then please simplify :)

Reply:
Ack, I'll remove or keep depending on how we start the server.